Wabbajack.Test/WorkQueueTests.cs

```diff
@@ -0,0 +1,273 @@
+            {
+                var subj = new BehaviorSubject<int>(Large);
+                var tcs = new TaskCompletionSource<bool>();
+                using (var queue = new WorkQueue(subj))
+                {
+                    Enumerable.Range(0, Large).Do(_ => queue.QueueTask(() => tcs.Task));
+                    // This call deadlocks, as the continuations is a WorkQueue while loop
+                    tcs.SetResult(true);
+                }
+            });
+            var completed = Task.WhenAny(Task.Delay(3000), task);
+            Assert.ReferenceEquals(completed, task);
+        }
+        #endregion
+
+        #region Known Parallel Work Collapse Pitfall
+        /// <summary>
+        /// Putting a single TCS completion source onto the WorkQueue will result in parallization collapse, where
+        /// all work is being done by one actual thread.  Similar to the deadlock scenario, this is just slightly different.
+        /// 
+        /// Since all worker tasks in charge of pulling off the queue were working on a single job driven by a single TCS,
+        /// when that TCS completes, the one thread that completed it is in charge of all the continuation.  All the continuation
+        /// tasks happen to be all Tasks in charge of pulling off the queue.  This results in one actual thread essentially calling a
+        /// Task.WhenAll() on all of our queue.Take tasks.  This means only one thread is now ping-ponging around doing the work, rather
+        /// than our desired number of threads working in parallel.
+        /// 
+        /// This will happen even if the WorkQueue is backed by Threads, rather than Task.Run() calls.  It's just the nature of how async
+        /// continuation is wired to work.
+        /// 
+        /// Other notes:
+        ///   This seems to fail when run in the normal pipeline of unit tests.  I think the timing gets interrupted by other tests?
+        ///   Disabled the test from being run automatically for now
+        /// 
+        /// TLDR:  Don't put the same work completion source to be done on the queue multiple times.
+        /// </summary>
+        public async Task ThreadCoalescenceExample()
+        {
+            var subj = new BehaviorSubject<int>(Large);
+            var tcs = new TaskCompletionSource<bool>();
+            object lockObj = new object();
+            using (var queue = new WorkQueue(subj))
+            {
+                Assert.AreEqual(Large, queue.DesiredNumWorkers);
+                Assert.AreEqual(Large, queue._tasks.Count);
+
+                bool[] workStartedArray = new bool[Large];
+                async Task Job(int num, bool[] b)
+                {
+                    // Mark work started as soon as job started
+                    lock (lockObj)
+                    {
+                        b[num] = true;
+                    }
+                    // Do lots of hard work for 1 second
+                    Thread.Sleep(5000);
```

Comment:
These two tests aren't really "tests", but rather testable concrete examples of pitfalls to watch out for when using WorkQueue.  Might find them interesting to look over